source/slang/slang-ast-dump.cpp

```diff
@@ -0,0 +1,615 @@
+            }
+            m_context->m_scopeWriteCount++;
+        }
+
+        ~ScopeWrite()
+        {
+            if (--m_context->m_scopeWriteCount == 0)
+            {
+                m_context->m_writer->emit(m_context->m_buf);
+            }
+        }
+
+        StringBuilder& getBuf() { return m_context->m_buf; }
+
+        operator StringBuilder&() { return m_context->m_buf; }
+
+        Context* m_context;
+    };
+
+    void dumpObject(const ReflectClassInfo& type, RefObject* obj);
+
+    void dumpObjectFull(const ReflectClassInfo& type, RefObject* obj, Index objIndex);
+    void dumpObjectReference(const ReflectClassInfo& type, RefObject* obj, Index objIndex);
+
+    void dump(NodeBase* node)
+    {
+        if (node == nullptr)
+        {
+            dumpPtr(nullptr);
+        }
+        else
+        {
+            dumpObject(node->getClassInfo(), node);
+        }
+    }
+
+    void dump(Substitutions* subs)
+    {
+        if (subs == nullptr)
+        {
+            dumpPtr(nullptr);
+        }
+        else
+        {
+            dumpObject(subs->getClassInfo(), subs);
+        }
+    }
+
+    void dump(const Name* name)
+    {
+        if (name == nullptr)
+        {
+            dumpPtr(nullptr);
+        }
+        else
+        {
+            dump(name->text);
+        }
+    }
+
+    void dump(const RefObject* obj)
+    {
+        if (obj == nullptr)
+        {
+            dumpPtr(nullptr);
+        }
+        else
+        {
+            // We don't know what this is!
+            ScopeWrite(this).getBuf() << "Unknown@" << size_t(obj);
+        }
+    }
+
+    template <typename T>
+    void dump(const List<T>& list)
+    {
+        m_writer->emit(" { \n");
+        m_writer->indent();
+        for (Index i = 0; i < list.getCount(); ++i)
+        {
+            dump(list[i]);
+            if (i < list.getCount() - 1)
+            {
+                m_writer->emit(",\n");
+            }
+            else
+            {
+                m_writer->emit("\n");
+            }
+        }
+        m_writer->dedent();
+        m_writer->emit("}");
+    }
+
+    void dump(SourceLoc sourceLoc)
+    {
+        SourceManager* manager = m_writer->getSourceManager();
+
+        {
+            ScopeWrite(this).getBuf() << "SourceLoc(" << sourceLoc.getRaw() << ")";
+        }
+
+        if (manager && sourceLoc.isValid())
+        {
+            HumaneSourceLoc humaneLoc = manager->getHumaneLoc(sourceLoc);
+            ScopeWrite(this).getBuf() << " " << humaneLoc.pathInfo.foundPath << ":" << humaneLoc.line;   
+        }
+    }
+
+    static char _getHexDigit(UInt v)
+    {
+        return (v < 10) ? char(v + '0') : char('a' + v - 10);
+    }
+
+    void dump(const UnownedStringSlice& slice)
+    {
+        m_writer->emitChar('\"');
+
+        {
+            ScopeWrite scope(this);
+            auto& buf = scope.getBuf();
+            for (const char c : slice)
+            {
+                if (c < 0x20 || c >= 0x80)
```

Comment:
Probably worth factoring into a helper subroutine just so this can be written as `if(!_isPrintable(c))` or similar.